addons/website/models/website_visitor.py

```diff
@@ -177,72 +199,82 @@ def _get_visitor_from_request(self, force_create=False):
         # This function can be called in json with mobile app.
         # In case of mobile app, no uid is set on the jsonRequest env.
         # In case of multi db, _env is None on request, and request.env unbound.
-        if not request:
+        if not (request and request.env and request.env.uid):
             return None
+
         Visitor = self.env['website.visitor'].sudo()
         visitor = Visitor
-        access_token = request.httprequest.cookies.get('visitor_uuid')
-        if access_token:
-            visitor = Visitor.with_context(active_test=False).search([('access_token', '=', access_token)])
-            # Prefetch access_token and other fields. Since access_token has a restricted group and we access
-            # a non restricted field (partner_id) first it is not fetched and will require an additional query to be retrieved.
-            visitor.access_token
+        access_token = self._get_access_token()
+
+        if force_create:
+            # This will create or return existing one
+            visitor_values = Visitor._get_visitor_create_values()
+            if len(str(access_token)) != 40:
+                visitor_values['partner_id'] = access_token
+            visitor_columns = ', '.join(visitor_values.keys())
+            visitor_template = ', '.join('%s' for _ in visitor_values)
+            query = f"""
+                INSERT INTO website_visitor ({visitor_columns})
+                VALUES ({visitor_template})
+                ON CONFLICT (access_token)
+                DO UPDATE SET
+                    last_connection_datetime='{visitor_values['last_connection_datetime']}',
+                    visit_count = CASE WHEN website_visitor.last_connection_datetime < NOW() AT TIME ZONE 'UTC' - INTERVAL '8 hours'
+                                       THEN website_visitor.visit_count + 1
+                                       ELSE website_visitor.visit_count
+                                  END
```

Comment:
This kind of update will be subject to concurrent update errors, especially as visitor IDs won't be unique anymore, I think.

Imagine a room full of people following a training, and asked to open the website.  Chances are that several of them will get the same visitor ID, and cause a concurrent update crash. You'll need to protect that with `SKIP LOCKED` or some equivalent.